src/gretel_trainer/relational/json.py

```diff
@@ -314,30 +314,48 @@ def _denormalize_json(
     table_name_mappings: dict[str, str],
     original_table_name: str,
 ) -> pd.DataFrame:
-    table_names = list(table_name_mappings.values())
+    # Throughout this function, "provenance name" refers to the json-breadcrumb-style name (e.g. foo^bar>baz)
+    # and "node name" refers to the table name as it appears in the RelationalData graph (e.g. foo_invented_{uuid})
+
+    # The provided `table_name_mappings` argument (from producer metadata) maps from provenance name to node name.
+    # `inverse_table_name_mappings` inverts that mapping (so: node name keys, provenance name values).
+    # `table_dict` replaces the keys (node namees) in `tables` with corresponding provenance names
+    table_node_names = list(table_name_mappings.values())
     inverse_table_name_mappings = {v: k for k, v in table_name_mappings.items()}
-    table_dict: dict = {inverse_table_name_mappings[k]: v for k, v in tables.items()}
-    for table_name in list(reversed(table_names)):
+    table_dict = {inverse_table_name_mappings[k]: v for k, v in tables.items()}
+
+    get_table = lambda node_name: _get_table(
+        rel_data, inverse_table_name_mappings, table_dict, node_name
+    )
+
+    for table_name in list(reversed(table_node_names)):
+        table_df = get_table(table_name)
         table_provenance_name = inverse_table_name_mappings[table_name]
-        empty_fallback = pd.DataFrame(
-            data={col: [] for col in rel_data.get_table_columns(table_name)},
-        )
-        table_df = table_dict.get(table_provenance_name, empty_fallback)
 
         if table_df.empty and _is_invented_child_table(table_name, rel_data):
-            p_name = rel_data.get_foreign_keys(table_name)[0].parent_table_name
-            parent_name = inverse_table_name_mappings[p_name]
+            # Mutate the parent dataframe by adding a column with empty lists
+
+            # We know invented child tables have exactly one foreign key
+            fk = rel_data.get_foreign_keys(table_name)[0]
+            parent_node_name = fk.parent_table_name
+            parent_provenance_name = inverse_table_name_mappings[parent_node_name]
             col_name = get_parent_column_name_from_child_table_name(
                 table_provenance_name
             )
-            kwargs = {col_name: table_dict[parent_name].apply(lambda x: [], axis=1)}
-            table_dict[parent_name] = table_dict[parent_name].assign(**kwargs)
+            parent_df = get_table(parent_node_name)
+            kwargs = {col_name: parent_df.apply(lambda x: [], axis=1)}
```

Comment:
what does this do? makes an empty list for every row?

Reply:
+1 to Matt's question and I think that `apply` and `assign` combo makes it a bit hard to understand.

Maybe writing it in this way would be easier to grasp?
```python
parent_df[col_name] = [[] for _ in range(len(parent_df))]
```

I think modifying `parent_df` is okay here, but if we didn't want to do that, we could
```python
parent_df.assign(**{col_name: [[] for _ in range(len(parent_df))]})
```

Reply:
Yeah this is definitely more clear, thanks. BTW, we _need_ to modify `parent_df` so that the column we add is present on subsequent iterations through the tables loop.